core/src/main/clojure/xtdb/compactor.clj

```diff
@@ -89,17 +89,24 @@
         (when (.isValid ev-ptr is-valid-ptr path)
           (.add merge-q {:ev-ptr ev-ptr, :row-copier row-copier})))
 
-      (loop []
+      (loop [previous-polygon? false]
         (when-let [{:keys [^EventRowPointer ev-ptr, ^RowCopier row-copier] :as q-obj} (.poll merge-q)]
-          (.copyRow row-copier (.getIndex ev-ptr))
-
-          (.writeLong recency-wtr
-                      (.getRecency ^IPolygonReader (calculate-polygon ev-ptr)))
-
-          (.nextIndex ev-ptr)
-          (when (.isValid ev-ptr is-valid-ptr path)
-            (.add merge-q q-obj))
-          (recur))))
+          (let [new-previous-polygon  (if-let [polygon (calculate-polygon ev-ptr)]
+                                        (do
+                                          (.copyRow row-copier (.getIndex ev-ptr))
+                                          (.writeLong recency-wtr (.getRecency ^IPolygonReader polygon))
+                                          true)
+
+                                        (do
+                                          ;; the first time we encounter an erase
+                                          (when previous-polygon?
+                                            (.copyRow row-copier (.getIndex ev-ptr))
+                                            (.writeLong recency-wtr -1))
+                                          false))]
```

Comment:
A bit of weird one. I could also check the `event-ptr` explicitly for `erase` if you think that is more readable.